### PR TITLE
Transform global derivatives of Jacobian to reference derivatives in `preprocessed_form`

### PR DESCRIPTION
**Pull request by jpdean**
This PR fixes https://github.com/FEniCS/ufl/issues/62 and https://github.com/FEniCS/ffcx/issues/373.

### CODE REVIEW

File: ufl/algorithms/apply_derivatives.py [hunk 1]
```diff
@@ -593,6 +591,25 @@ def _grad(self, o):
         # 1) n = count number of Grads, get f
         # 2) if not f.has_derivatives(n): return zero(...)
 
+    def _grad_to_reference_grad(self, o, K):
```
Comment (wence-) [inline ufl/algorithms/apply_derivatives.py:594]: Nitpick, this function doesn't need `self` I think, so could it just be a free function?

File: ufl/algorithms/apply_derivatives.py [hunk 2]
```diff
@@ -491,24 +491,26 @@ def __init__(self, geometric_dimension):
...
-        """Default for geometric quantities is dg/dx = 0 if piecewise constant, otherwise keep Grad(g).
+        """Default for geometric quantities is do/dx = 0 if piecewise constant,
+        otherwise transform derivatives to reference derivatives.
         Override for specific types if other behaviour is needed."""
         if is_cellwise_constant(o):
             return self.independent_terminal(o)
         else:
             # TODO: Which types does this involve? I don't think the
             # form compilers will handle this.
-            return Grad(o)
```
Comment (wence-) [inline ufl/algorithms/apply_derivatives.py:501]: I think you can also remove the `TODO` comment?

**Review by wence- — COMMENTED** [review]
Thanks! Modulo minor comments I think this looks right.

File: ufl/algorithms/apply_derivatives.py (hunk 1 above)
Comment (jpdean) [reply ufl/algorithms/apply_derivatives.py:594]: Sure! I put it there because it is only used by the `GradRuleset` class, but I'm happy to move it

**Review by garth-wells — APPROVED** [review]

**Pull request merged**